ethology/annotations/validators.py

```diff
@@ -0,0 +1,192 @@
+            "shape_attributes": ["x", "y", "width", "height"],
+        },
+        init=False,
+    )
+
+    # Note: the validators are applied in order
+    @path.validator
+    def _file_is_json(self, attribute, value):
+        _check_file_is_json(value)
+
+    @path.validator
+    def _file_matches_JSON_schema(self, attribute, value):
+        _check_file_matches_schema(value, self.schema)
+
+    @path.validator
+    def _file_contains_required_keys(self, attribute, value):
+        """Ensure that the VIA JSON file contains the required keys."""
+        # Read data as dict
+        with open(value) as file:
+            data = json.load(file)
+
+        # Check first level keys
+        _check_required_keys_in_dict(self.required_keys["main"], data)
+
+        # Check keys in nested dicts
+        for img_str, img_dict in data["_via_img_metadata"].items():
+            # Check keys for each image dictionary
+            _check_required_keys_in_dict(
+                self.required_keys["images"],
+                img_dict,
+                additional_message=f" for {img_str}",
+            )
+
+            # Check keys for each region in an image
+            for i, region in enumerate(img_dict["regions"]):
+                # Check keys under first level per region
+                _check_required_keys_in_dict(
+                    self.required_keys["regions"],
+                    region,
+                    additional_message=f" for region {i} under {img_str}",
+                )
+
+                # Check keys under "shape_attributes" per region
+                _check_required_keys_in_dict(
+                    self.required_keys["shape_attributes"],
+                    region["shape_attributes"],
+                    additional_message=f" for region {i} under {img_str}",
+                )
+
+
+@define
+class ValidCOCO:
+    """Class for valid COCO JSON files.
+
+    It checks the input file is a valid JSON file, matches
+    the COCO schema and contains the required keys.
+
+    Attributes
+    ----------
+    path : pathlib.Path
+        Path to the COCO JSON file, passed as an input.
+    schema : dict
+        The JSON schema is set to the default COCO schema.
+    required_keys : dict
+        The required keys for the COCO JSON file.
+
+    Raises
+    ------
+    ValueError
+        If the JSON file cannot be decoded.
+    jsonschema.exceptions.ValidationError
+        If the type of any of the keys in the JSON file
+        does not match the type specified in the schema.
+    jsonschema.exceptions.SchemaError
+        If the schema is invalid.
+    ValueError
+        If the COCO JSON file is missing any of the required keys.
+
+    """
+
+    path: Path = field()
+    schema: dict = field(
+        default=_get_default_COCO_schema(),
+        init=False,
+        # init=False makes the attribute to be unconditionally initialized
+        # with the specified default
```

Comment:
I'm not sure what that comment means. Does it mean I cannot overwrite the schema passing a different dict?
In any case, this comment should be moved to the first occurrence of `init=False` in this file.